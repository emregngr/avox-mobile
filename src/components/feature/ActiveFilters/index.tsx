import { MaterialCommunityIcons } from '@expo/vector-icons'
import React, { memo, useCallback, useMemo, useRef } from 'react'
import { ScrollView, TouchableOpacity, View } from 'react-native'

import { ThemedText } from '@/components/common/ThemedText'
import { getLocale } from '@/locales/i18next'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'
import { getActiveFilterLabel } from '@/utils/feature/getActiveFilterLabel'

interface ActiveFiltersProps {
  filters: any
  onClearAll: () => void
  onRemove: (key: string) => void
}

const AUTO_SCROLL_DELAY = 300

export const ActiveFilters = memo(({ filters, onClearAll, onRemove }: ActiveFiltersProps) => {
  const { selectedTheme } = useThemeStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const getActiveFilterLabelValue = getActiveFilterLabel()

  const scrollViewRef = useRef<ScrollView>(null)

  const getFilterCount = useCallback(() => {
    let count = 0
    if (!filters) return 0
    Object?.entries(filters)?.forEach(([key, value]) => {
      if (Array?.isArray(value)) {
        count += value?.length
      } else if (value !== null && value !== undefined && value !== false) {
        count += 1
      }
    })
    return count
  }, [filters])

  const filterCount = useMemo(() => getFilterCount(), [getFilterCount])

  const filterEntries = useMemo(
    () =>
      Object.entries(filters).filter(
        ([_, value]) => value !== null && value !== undefined && value !== false,
      ),
    [filters],
  )

  const handleRemove = useCallback(
    (key: string) => {
      const elementIndex = filterEntries.findIndex(([k, _]) => k === key)
      const isInLastThree = elementIndex >= filterEntries?.length - 3

      onRemove(key)

      if (isInLastThree) {
        setTimeout(() => {
          scrollViewRef?.current?.scrollToEnd({ animated: true })
        }, AUTO_SCROLL_DELAY)
      }
    },
    [filterEntries, onRemove],
  )

  if (filterCount === 0) return null

  return (
    <View className="mr-12">
      <ScrollView
        contentContainerClassName="items-center"
        ref={scrollViewRef}
        showsHorizontalScrollIndicator={false}
        horizontal
      >
        <View className="flex-row items-center pr-4">
          <TouchableOpacity
            activeOpacity={0.7}
            className="bg-error px-3 py-1.5 rounded-full overflow-hidden mr-8 left-6"
            hitSlop={10}
            onPress={onClearAll}
            testID="active-filters-clear-button"
          >
            <ThemedText color="text-100" type="body3">
              {getLocale('clear')} ({filterCount})
            </ThemedText>
          </TouchableOpacity>

          {filterEntries.map(([key, filterValue]) => (
            <View
              className="bg-secondary-100 px-3 py-1.5 rounded-full overflow-hidden mr-2 flex-row items-center"
              key={key}
            >
              <ThemedText color="text-100" type="body3">
                {getActiveFilterLabelValue(key, filterValue)}
              </ThemedText>
              <TouchableOpacity
                activeOpacity={0.7}
                className="bg-background-quaternary rounded-full overflow-hidden ml-2"
                hitSlop={10}
                onPress={() => handleRemove(key)}
                testID={`active-filters-remove-${key}`}
              >
                <MaterialCommunityIcons color={colors?.onPrimary100} name="close" size={14} />
              </TouchableOpacity>
            </View>
          ))}
        </View>
      </ScrollView>
    </View>
  )
})
