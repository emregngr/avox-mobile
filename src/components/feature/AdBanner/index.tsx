import React, { memo, useState } from 'react'
import { View } from 'react-native'
import { BannerAd, BannerAdSize } from 'react-native-google-mobile-ads'
import { useSafeAreaInsets } from 'react-native-safe-area-context'

import { isProduction } from '@/config/env/environment'
import { AD_KEYWORDS } from '@/constants/adKeywords'
import { cn } from '@/utils/common/cn'

interface AdBannerProps {
  adUnitId: string
}

export const AdBanner = memo(({ adUnitId }: AdBannerProps) => {
  const { bottom } = useSafeAreaInsets()

  const [isAdLoaded, setIsAdLoaded] = useState<boolean>(false)

  const requestOptions = {
    keywords: AD_KEYWORDS,
    requestNonPersonalizedAdsOnly: true,
  }

  const handleAdLoaded = () => {
    setIsAdLoaded(true)
  }

  const handleAdFailedToLoad = () => {
    setIsAdLoaded(false)
  }

  return !__DEV__ && isProduction() ? (
    <View
      className={cn('bg-background-primary', isAdLoaded ? 'h-auto' : 'h-0')}
      style={{ paddingBottom: bottom }}
      testID="ad-banner-container"
    >
      <BannerAd
        onAdFailedToLoad={handleAdFailedToLoad}
        onAdLoaded={handleAdLoaded}
        requestOptions={requestOptions}
        size={BannerAdSize.ANCHORED_ADAPTIVE_BANNER}
        unitId={adUnitId}
      />
    </View>
  ) : null
})

AdBanner.displayName = 'AdBanner'
