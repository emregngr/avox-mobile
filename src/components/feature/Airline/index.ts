export * from './AirlineCard'
export * from './AirlineDetail'
export * from './AirlineFavoriteItemCard'
export * from './AirlinesTab'
export * from './SkeletonAirlineCard'
