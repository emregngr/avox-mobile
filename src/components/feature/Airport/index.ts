export * from './AirportCard'
export * from './AirportCardSkeleton'
export * from './AirportDetail'
export * from './AirportsTab'
