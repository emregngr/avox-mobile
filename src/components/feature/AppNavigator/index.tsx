import { Stack } from 'expo-router'
import { useMemo } from 'react'

import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'

interface AppNavigatorProps {
  isAppReady: boolean | null
}

export const AppNavigator = ({ isAppReady }: AppNavigatorProps) => {
  const { selectedTheme } = useThemeStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const stackScreenOptions = useMemo(
    () => ({
      animation: 'slide_from_bottom' as const,
      contentStyle: { backgroundColor: colors?.background?.blur },
      headerShown: false,
    }),
    [colors],
  )

  if (!isAppReady) {
    return (
      <Stack initialRouteName="index" screenOptions={stackScreenOptions}>
        <Stack.Screen name="index" />
      </Stack>
    )
  }

  return (
    <Stack initialRouteName="index" screenOptions={stackScreenOptions}>
      <Stack.Screen name="index" />
      <Stack.Screen name="(auth)" />
      <Stack.Screen name="(tabs)" />
      <Stack.Screen name="(onboarding)" />
      <Stack.Screen name="(maintenance)" />
      <Stack.Screen name="(force-update)" />
      <Stack.Screen name="(account)" />
      <Stack.Screen name="(settings)" />
      <Stack.Screen name="(airline)" />
      <Stack.Screen name="(airport)" />
      <Stack.Screen name="(breaking-news)" />
      <Stack.Screen name="(destination)" />
      <Stack.Screen name="(airplane)" />
      <Stack.Screen name="token-expire" />
      <Stack.Screen name="(web-view-modal)" options={{ presentation: 'modal' }} />
      <Stack.Screen name="(image-modal)" options={{ presentation: 'modal' }} />
    </Stack>
  )
}
