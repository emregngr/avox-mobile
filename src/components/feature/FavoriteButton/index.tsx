import { Ionicons } from '@expo/vector-icons'
import React, { memo, useMemo } from 'react'
import { ActivityIndicator, TouchableOpacity } from 'react-native'

import { useFavoriteToggle } from '@/hooks/services/useFavoriteToggle'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'
import type { FavoriteItem } from '@/types/feature/favorite'

const FavoriteButtonComponent = ({ id, type }: FavoriteItem) => {
  const { selectedTheme } = useThemeStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const { handleFavoritePress, isFavorite, isPending } = useFavoriteToggle({ id, type })

  return (
    <TouchableOpacity
      activeOpacity={0.7}
      className="bg-background-quaternary p-2 rounded-full overflow-hidden absolute top-2 right-2"
      disabled={isPending}
      hitSlop={20}
      onPress={handleFavoritePress}
    >
      {isPending ? (
        <ActivityIndicator color={colors?.tertiary100} size="small" />
      ) : (
        <Ionicons
          color={isFavorite ? colors?.tertiary100 : colors?.onPrimary100}
          name={isFavorite ? 'heart' : 'heart-outline'}
          size={20}
        />
      )}
    </TouchableOpacity>
  )
}

export const FavoriteButton = memo(FavoriteButtonComponent)
