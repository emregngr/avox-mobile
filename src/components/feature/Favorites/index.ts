export * from './FavoriteAirlinesList'
export * from './FavoriteAirportsList'
