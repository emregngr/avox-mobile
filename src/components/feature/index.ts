export * from './ActiveFilters'
export * from './Airline'
export * from './Airport'
export * from './AppNavigator'
export * from './FavoriteButton'
export * from './Favorites'
export * from './FilterModal'
export * from './Home'
