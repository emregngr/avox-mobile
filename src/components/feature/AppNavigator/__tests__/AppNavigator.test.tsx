import { render } from '@testing-library/react-native'
import type { ReactNode } from 'react'
import React from 'react'

import { AppNavigator } from '@/components/feature/AppNavigator'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'

jest.mock('@/store/theme')

const mockedUseThemeStore = useThemeStore as jest.MockedFunction<typeof useThemeStore>

jest.mock('expo-router', () => {
  const { View } = require('react-native')

  const MockedStack = ({
    children,
    screenOptions,
  }: {
    children: ReactNode
    screenOptions: any
  }) => (
    <View data-screenOptions={screenOptions} testID="MockStack">
      {children}
    </View>
  )

  const MockScreen = (props: { name: string }) => <View testID={`MockStackScreen-${props.name}`} />

  MockedStack.Screen = MockScreen

  return {
    Stack: MockedStack,
  }
})

beforeEach(() => {
  mockedUseThemeStore.mockReturnValue({ selectedTheme: 'light' })
})

describe('AppNavigator Component', () => {
  it('should render only the "index" screen when isAppReady is false', () => {
    const { getByTestId, queryByTestId } = render(<AppNavigator isAppReady={false} />)

    expect(getByTestId('MockStackScreen-index')).toBeTruthy()
    expect(queryByTestId('MockStackScreen-(auth)')).toBeNull()
    expect(queryByTestId('MockStackScreen-(tabs)')).toBeNull()
  })

  it('should render all screens when isAppReady is true', () => {
    const { getByTestId } = render(<AppNavigator isAppReady />)

    expect(getByTestId('MockStackScreen-index')).toBeTruthy()
    expect(getByTestId('MockStackScreen-(auth)')).toBeTruthy()
    expect(getByTestId('MockStackScreen-(tabs)')).toBeTruthy()
    expect(getByTestId('MockStackScreen-(onboarding)')).toBeTruthy()
    expect(getByTestId('MockStackScreen-(maintenance)')).toBeTruthy()
    expect(getByTestId('MockStackScreen-(force-update)')).toBeTruthy()
    expect(getByTestId('MockStackScreen-(account)')).toBeTruthy()
    expect(getByTestId('MockStackScreen-(settings)')).toBeTruthy()
    expect(getByTestId('MockStackScreen-(airline)')).toBeTruthy()
    expect(getByTestId('MockStackScreen-(airport)')).toBeTruthy()
    expect(getByTestId('MockStackScreen-(breaking-news)')).toBeTruthy()
    expect(getByTestId('MockStackScreen-(destination)')).toBeTruthy()
    expect(getByTestId('MockStackScreen-(airplane)')).toBeTruthy()
    expect(getByTestId('MockStackScreen-token-expire')).toBeTruthy()
    expect(getByTestId('MockStackScreen-(web-view-modal)')).toBeTruthy()
    expect(getByTestId('MockStackScreen-(image-modal)')).toBeTruthy()
  })

  it('should pass the correct screenOptions prop to the Stack component based on the selected theme', () => {
    mockedUseThemeStore.mockReturnValue({ selectedTheme: 'dark' })

    const expectedBackgroundColor = themeColors.dark.background.blur

    const { getByTestId } = render(<AppNavigator isAppReady />)

    const stack = getByTestId('MockStack')
    const passedOptions = stack.props['data-screenOptions']

    expect(passedOptions.contentStyle.backgroundColor).toBe(expectedBackgroundColor)
    expect(passedOptions.animation).toBe('slide_from_bottom')
    expect(passedOptions.headerShown).toBe(false)
  })
})

describe('AppNavigator Component Snapshot', () => {
  it('should render the AppNavigator Component successfully', () => {
    const { toJSON } = render(<AppNavigator isAppReady />)

    expect(toJSON()).toMatchSnapshot()
  })
})
