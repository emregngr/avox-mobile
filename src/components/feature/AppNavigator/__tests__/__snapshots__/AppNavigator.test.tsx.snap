// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AppNavigator Component Snapshot should render the AppNavigator Component successfully 1`] = `
<View
  data-screenOptions={
    {
      "animation": "slide_from_bottom",
      "contentStyle": {
        "backgroundColor": "#FFFFFFCC",
      },
      "headerShown": false,
    }
  }
  testID="MockStack"
>
  <View
    testID="MockStackScreen-index"
  />
  <View
    testID="MockStackScreen-(auth)"
  />
  <View
    testID="MockStackScreen-(tabs)"
  />
  <View
    testID="MockStackScreen-(onboarding)"
  />
  <View
    testID="MockStackScreen-(maintenance)"
  />
  <View
    testID="MockStackScreen-(force-update)"
  />
  <View
    testID="MockStackScreen-(account)"
  />
  <View
    testID="MockStackScreen-(settings)"
  />
  <View
    testID="MockStackScreen-(airline)"
  />
  <View
    testID="MockStackScreen-(airport)"
  />
  <View
    testID="MockStackScreen-(breaking-news)"
  />
  <View
    testID="MockStackScreen-(destination)"
  />
  <View
    testID="MockStackScreen-(airplane)"
  />
  <View
    testID="MockStackScreen-token-expire"
  />
  <View
    testID="MockStackScreen-(web-view-modal)"
  />
  <View
    testID="MockStackScreen-(image-modal)"
  />
</View>
`;
